tempesta_fw/http.c

```diff
@@ -4632,30 +4765,43 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, const TfwFsmData *data)
 				"Request parsing inconsistency");
 			return TFW_BLOCK;
 		}
-		r = tfw_gfsm_move(&conn->state, TFW_HTTP_FSM_REQ_CHUNK,
-				  &data_up);
-		T_DBG3("TFW_HTTP_FSM_REQ_CHUNK return code %d\n", r);
-		if (r == TFW_BLOCK) {
-			TFW_INC_STAT_BH(clnt.msgs_filtout);
-			tfw_http_req_parse_block(req, 403,
-						 "postponed request has been"
-						 " filtered out");
-			return TFW_BLOCK;
+		if (TFW_MSG_H2(req) && tfw_h2_stream_req_complete(req->stream)) {
+			if (tfw_h2_parse_req_finish(req)) {
```

Comment:
We need this check here, because of all that skb fragment processing effects inside of `ss_skb_process()`.

How do we process h2 traffic:
- First we try to find frame borders in `tfw_h2_frame_process()`. We use  `ss_skb_process()`  to walk among skb fragments and split skb at frame end. At the end of this stage we have skb which contain only h2 headers. Let's assume that the headers is fully received, no body required for that request and the skb is fragmented. So the `STREAM_END`  flag will be received there.
- Since the `STREAM_END` flag is received, we mark the stream as finished to indicate request end for parser layer.
- Then the fragmented skb is passed into `tfw_http_req_process()` to parse the request. With `ss_skb_process()` we call `tfw_h2_parse_req()` _for each_ skb fragment.

Since the `tfw_h2_parse_req()` is called for every skb fragment, we can't have `tfw_h2_parse_req_finish()` inside it. The stream is _already_ marked as finished and after the very first fragment was parsed, the `tfw_h2_parse_req_finish()` would be called. The remaining fragments will remain unparsed and glued with the parsed ones.

I understand your wish to hide all the parser routines (and h2 specific processing) under single `actor` for `ss_skb_process()`. But since in h2  _the framing_ is done on framing level, not on parser one, we either need to explicitly call a parser function for the message end (in current patch), or make it implicitly by some extra state sharing.

I decided that the next parser API looks a bit cleaner:
- one function that parses a message from contiguous buffer - `tfw_h2_parse_req()`. It never return `T_OK` on success, only `T_POSTPONE`, since in h2 _the framing_ is done on framing level, not on parser one. 
- one function to explicitly indicate the message end and fixup the parser routines - `tfw_h2_parse_req_finish()`

Reply:
Yeah, that's about skb fragments - I forgot about them. Thanks!